CoastalResilienceNetwork/GeositeFramework
# Create banch to upgrade to Esri ArcGIS JS API v4.2

## Overview

This PR begins the work of creating a new branch to test out upgrading the Esri ArcGIS JavaScript API from v3.x to v4.2. I've already created a new branch off `prototype` at `js-api-4x-pilot`, and this branch starts the process of upgrading the framework to accommodate the new API.

I tried to have each commit here address only one or two changes for the upgrade. Some of the adjustments include:

- updating names or import paths to correspond to the new names in the v4.2 API
- commenting out (and writing searchable text) for features which aren't or aren't yet supported
- setting up a new basemap, along with the related CSS rules to mount it and adjust how the basemap layer selector works
- splitting the `Map` and `MapView` in some cases to match the new API. The new API uses the 'Map' as a store for data, and the 'MapView' as the mechanism for handling the UI. Here I've done some preliminary work to split those, but we'll probably need up update the model more thoroughly to handle it.
- adjusting the `MapWrapper`'s calls to toggle layers to use the API's new method names

## Screenshots

The app currently looks like this, and includes a working basemap selector along with partially-working layers in the `regional-planning` plugin (from a PR here: https://github.com/CoastalResilienceNetwork/regional-planning/pull/75)

<img width="1024" alt="screen shot 2017-01-11 at 5 48 39 pm" src="https://cloud.githubusercontent.com/assets/4165523/21870297/a9af9248-d829-11e6-846e-9bdd70529d47.png">

<img width="1030" alt="screen shot 2017-01-11 at 5 48 51 pm" src="https://cloud.githubusercontent.com/assets/4165523/21870299/aed69758-d829-11e6-8f18-aa6917e9e148.png">

<img width="1032" alt="screen shot 2017-01-11 at 5 49 36 pm" src="https://cloud.githubusercontent.com/assets/4165523/21870302/b21b2ef6-d829-11e6-9b63-55ce366651a3.png">

<img width="1028" alt="screen shot 2017-01-11 at 5 49 51 pm" src="https://cloud.githubusercontent.com/assets/4165523/21870307/b608a7e6-d829-11e6-854a-64a78c1d8fe5.png">

## Notes

Only some of the layers from `regional-planning` are working thus far, including the Inundation Depth tile layer and some of the Hurricane Sandy Layers. The other layers are making requests to the map services when added, but they're currently grabbing the incorrect sublayers. Failed layers seem to try for index `-1`, and the Hurricane Sandy County impact seems to be grabbing sublayers from indices `[0, 2]`. 

I'm inclined to make another card to handle this, as I'm not certain yet whether it needs to be handled at the level of the plugin or at the level of the framework.

I've catalogued a bunch of additional notes here:

https://github.com/CoastalResilienceNetwork/GeositeFramework/issues/828

and there's a PR which needs to be reviewed in tandem with this one here:

https://github.com/CoastalResilienceNetwork/regional-planning/pull/75

## Testing
- get this branch along with the similar branch from https://github.com/CoastalResilienceNetwork/regional-planning/pull/75
- rebuild in Visual Studio and open in the browser
- verify that you see a map and that the plugins load without error
- select different basemaps from the dropdown and verify that they render
- toggle `regional-planning` on and select the Inundation Layers and the Hurricane Sandy Layers and verify that they render on the map
- toggle the other layers and verify that you see the requests being made successfully -- even if the images are currently empty

Connects #828
